File: src/javascript/algorithms/array/minimumSizeSubarraySum/minSizeSubarraySum.js
// Sliding window: TC: O(n), SC: O(1)

function minSizeSubarraySum(nums, target) {
    
    let left = right = total = 0;
    let result = Infinity;

    while(right < nums.length) {
        total += nums[right];
        while(total >= target) {
            result = Math.min(right - left + 1, result);
            total -= nums[left++];
        }
        right++;
    }

    return result === Infinity ? 0 : result;
}

let target1 = 7, nums1 = [2,4,1,2,4,3];
let target2 = 5, nums2 = [1, 5, 5, 5];
let target3 = 15, nums3 = [2, 2, 2, 2, 2];

console.log(minSizeSubarraySum(nums1, target1));
console.log(minSizeSubarraySum(nums2, target2));
console.log(minSizeSubarraySum(nums3, target3));

File: test/arrays.py/min_size_subarray_sum.py
def min_size_subarray_sum(nums, target):
    left = right = 0
    total = 0
    min_length = float("inf")

    for right in range(len(nums)):
        total += nums[right]

        while total >= target:
            min_length = min(min_length, right-left+1)
            total -= nums[left]
            left += 1
    return min_length if min_length != float("inf") else 0

# Test cases
if __name__ == "__main__":
    test_cases = [
        {"target": 7, "nums": [2, 4, 1, 2, 4, 3]},  # Output: 2
        {"target": 5, "nums": [1, 5, 5, 5]},        # Output: 1
        {"target": 15, "nums": [2, 2, 2, 2, 2]},    # Output: 0
    ]
    
    for test in test_cases:
        print(f"Input: target = {test['target']}, nums = {test['nums']}")
        print(f"Output: {min_size_subarray_sum(test['nums'], test['target'])}")
        print()
